winch/codegen/src/codegen/mod.rs

```diff
@@ -113,16 +186,38 @@ where
                 $(
                     fn $visit(&mut self $($(,$arg: $argty)*)?) -> Self::Output {
                         self.0.$visit($($($arg.clone()),*)?)?;
-                        Ok(self.1.$visit($($($arg),*)?))
+                        // Only visit operators if the compiler is in a reachable code state. If
+                        // the compiler is in an unrechable code state, most of the operators are
+                        // ignored except for If, Block, Loop, Else and End. These operators need
+                        // to be observed in order to keep the control stack frames balanced and to
+                        // determine if reachability should be restored.
+                        if self.1.is_reachable() || matches!(stringify!($op), "If" | "Block" | "Loop" | "Else" | "End") {
```

Comment:
Will this do a string-compare at runtime? I think it would be better to have a predicate on the opcode (actual enum value) that returns a bool -- call it something like `visit_op_when_unreachable` -- and use it here.

Reply:
Good call, updated!